From 9550a3120dfe226584b8126e27bae0f84ebd9a8f Mon Sep 17 00:00:00 2001
From: Roberto Perpuly <robertop2004@gmail.com>
Date: Tue, 24 Mar 2015 15:23:32 -0700
Subject: [PATCH] fix crash bug in wxPropertyGrid

Fix crash that ocurred when trying to set boolean values to be rendered as
checkboxes using wxPropertyGridManager::SetPropertyAttributeAll

See http://trac.wxwidgets.org/ticket/16922
---
 samples/propgrid/propgrid.cpp | 11 +++++++++++
 samples/propgrid/propgrid.h   |  2 ++
 src/propgrid/manager.cpp      |  2 +-
 3 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/samples/propgrid/propgrid.cpp b/samples/propgrid/propgrid.cpp
index bef0fc4..5d17d63 100644
--- a/samples/propgrid/propgrid.cpp
+++ b/samples/propgrid/propgrid.cpp
@@ -384,6 +384,7 @@ enum
     ID_ENABLE,
     ID_SETREADONLY,
     ID_HIDE,
+    ID_TOGGLE_RENDER_CHECKBOX,
     ID_DELETE,
     ID_DELETER,
     ID_DELETEALL,
@@ -492,6 +493,7 @@ wxBEGIN_EVENT_TABLE(FormMain, wxFrame)
     EVT_MENU( ID_ENABLE, FormMain::OnEnableDisable )
     EVT_MENU( ID_SETREADONLY, FormMain::OnSetReadOnly )
     EVT_MENU( ID_HIDE, FormMain::OnHide )
+    EVT_MENU( ID_TOGGLE_RENDER_CHECKBOX, FormMain::OnToggleRenderCheckbox )
 
     EVT_MENU( ID_ITERATE1, FormMain::OnIterate1Click )
     EVT_MENU( ID_ITERATE2, FormMain::OnIterate2Click )
@@ -1968,6 +1970,7 @@ FormMain::FormMain(const wxString& title, const wxPoint& pos, const wxSize& size
 
     m_propGrid = NULL;
     m_panel = NULL;
+    m_renderCheckbox = false;
 
 #ifdef __WXMAC__
     // we need this in order to allow the about menu relocation, since ABOUT is
@@ -2102,6 +2105,8 @@ FormMain::FormMain(const wxString& title, const wxPoint& pos, const wxSize& size
     menuTry->AppendSeparator();
     menuTry->AppendCheckItem(ID_STATICLAYOUT, wxT("Static Layout"),
         wxT("Switches between user-modifiable and static layouts.") );
+    menuTry->AppendCheckItem(ID_TOGGLE_RENDER_CHECKBOX, wxT("Render boolean values as checkboxes"),
+        wxT("Renders boolean values as checkboxes"));
     menuTry->Append(ID_SETCOLUMNS, wxT("Set Number of Columns") );
     menuTry->AppendSeparator();
     menuTry->Append(ID_TESTXRC, wxT("Display XRC sample") );
@@ -2532,6 +2537,12 @@ void FormMain::OnHide( wxCommandEvent& WXUNUSED(event) )
     m_pPropGridManager->HideProperty( id, true );
 }
 
+void FormMain::OnToggleRenderCheckbox( wxCommandEvent& WXUNUSED(event))
+{
+    m_renderCheckbox = !m_renderCheckbox;
+    m_pPropGridManager->SetPropertyAttributeAll( wxPG_BOOL_USE_CHECKBOX, m_renderCheckbox);
+}
+
 // -----------------------------------------------------------------------
 
 #include "wx/colordlg.h"
diff --git a/samples/propgrid/propgrid.h b/samples/propgrid/propgrid.h
index b017e0e..359c120 100644
--- a/samples/propgrid/propgrid.h
+++ b/samples/propgrid/propgrid.h
@@ -154,6 +154,7 @@ public:
 
     wxString        m_savedState;
 
+    bool            m_renderCheckbox;
 
     void CreateGrid( int style, int extraStyle );
     void FinalizeFramePosition();
@@ -186,6 +187,7 @@ public:
     void OnEnableDisable( wxCommandEvent& event );
     void OnSetReadOnly( wxCommandEvent& event );
     void OnHide( wxCommandEvent& event );
+    void OnToggleRenderCheckbox( wxCommandEvent& event );
     void OnSetBackgroundColour( wxCommandEvent& event );
     void OnClearModifyStatusClick( wxCommandEvent& event );
     void OnFreezeClick( wxCommandEvent& event );
diff --git a/src/propgrid/manager.cpp b/src/propgrid/manager.cpp
index 019530e..84b3481 100644
--- a/src/propgrid/manager.cpp
+++ b/src/propgrid/manager.cpp
@@ -1435,7 +1435,7 @@ void wxPropertyGridManager::RefreshProperty( wxPGProperty* p )
 {
     wxPropertyGrid* grid = p->GetGrid();
 
-    if ( GetPage(m_selPage)->GetStatePtr() == p->GetParent()->GetParentState() )
+    if ( p->GetParent() && GetPage(m_selPage)->GetStatePtr() == p->GetParent()->GetParentState() )
         grid->RefreshProperty(p);
 }
 
-- 
1.9.5 (Apple Git-50.3)

